### DISCUSSION TITLE
vscode + jupyter note book - pylance not recognizing local code python path

### DISCUSSION THREAD

**Discussion by dss010101** [microsoft/pylance-release#3460] [Enhancement]
I am not sure if this is an issue or a configuration problem.  I am developing against a local virtual environment in visual studio code.
I often use jupyter for quickly testing things and recently have started doing so within vscode itself.  i would basically open a notebook, split into the right window and connect to the same virtual environment i am developing in.

After doing this - i notice 'squiggly' lines start appearing where i have imports for code i have written. 
It seems with jupyter note book opened in visual studio, pylance is not recognizing my source code paths.
Interestingly, it is still finding all of the libraries installed in the virtual environment.   To fix this, i have to close the jupyter notebook and restart vscode.

anyone else experience this?

**Comment by judej** [comment]
Thanks for the report. Since the notebook and the `.py` file are in two different workspaces, the changes in `.py` will only be picked up after the files has been saved. Further, there will be a delay before the changes are picked up.

I will add this to discussions as a potential enhancement.

**Comment by dss010101** [comment]
note - this is an issue with changes not being picked up or recognized, it's simply whenever i open on jupyter notebox in vscode agains the same virtual environment im developing/debugging against - my local code is not recognized by pylance it would seem.  I can run and debug fine..those those annoying squiggly lines appear everywhere for just my code base.  so the python path to my project is getting lost somewhere...

**Reply by debonte** [reply]
@msingh00, can you provide a sample that reproduces this problem? Importing types from `.py` files in the same workspace as my notebook works fine for me.

**Reply by dss010101** [reply]
The issue isnt importing types from .py into a notebook.  To best describe...

Below is an image of the import section of a python file before opening my notebook in a separate tab:

![image](https://user-images.githubusercontent.com/855627/200243326-2dbbc44f-6e05-4f62-8a96-f97179e29b0e.png)

After i open the notebook this is what it looks like:

![image](https://user-images.githubusercontent.com/855627/200244358-981075e2-8f2b-415b-8933-d31097a3b133.png)

You can see that pylance seems to lose the ability to recognize some imports. i can still run and debug fine - just that i dont get the context sensitive editing help any longer until i close the notebook, and restart vscode.
Those particular imports are code i have written, as opposed to libraries that have been pip-installed into the virtual environment.  So it must have to do something with my workspace's python path getting screwed up when the notebook opens (at least for pylance - as mentioned the code will run fine and can be debugged).
Both my code and the notebook are using the same virtual environment.

**Reply by dss010101** [reply]
curious if you have any further thoughts on this.  my suspicion is that when i open a notebook in vscode, it some how wipes or changes my phython path for vscode, but im not sure.

I do see that when i open a note back and print out the PATH environment variable, it lists the path to my virtual environment's script directory twice(duplicated), where as if i did the same in vs code, it will show that path only once.

i have my code's python path defined in a .env file

**Reply by debonte** [reply]
@msingh00, I've been unable to repro this behavior. Can you provide sample code that reproduces this? If so, please file an issue with repro steps.

**Reply by dss010101** [reply]
i dont think the issue is code.  i will when time permit see if i can create a smaller project that can replicates this.
in the mean time, any thoughts as to what may affect the python path of vscode editor when the jupyter notebook is opened in the same workspace?
as a start is there anythingi  can read to understand how pylance searches for python path? in my project, i have them defined in a .env file.   that env file is referenced from the workspace.code-workspace file as such:
```
		"terminal.integrated.env.windows": {
			"python.envFile": "${workspaceFolder}/.env",
			//"PYTHONPATH": "${workspaceFolder}/src}"
		},
		"python.defaultInterpreterPath": "${env:PYTHON_INSTALL_LOC}",
		"typescript.tsdk": "node_modules/typescript/lib",
		"window.zoomLevel": -1,
		"files.exclude": {
			"**/__pycache__": true
		},
```
the env file looks like this:
```
PYTHONPATH=./src;./src/data;./src/studies;./src/utils;
PYTHON_INSTALL_LOC=./.venv/Scripts/python
TMP_DIR = c:/temp/
```
